src/sage/rings/species.py

```diff
@@ -0,0 +1,1468 @@
+            {((1,2,3,4)(5,6)(7,8)(9,10),): (frozenset({8, 5, 6, 7}), frozenset({1, 2, 3, 4, 9, 10}))}
+        """
+        return "{" + f"{self._dis}: {self._dompart}" + "}"
+
+
+class AtomicSpecies(UniqueRepresentation, Parent, ElementCache):
+    @staticmethod
+    def __classcall__(cls, names):
+        """
+        Normalize the arguments.
+
+        TESTS::
+
+            sage: A1 = AtomicSpecies("X")
+            sage: A2 = AtomicSpecies("Y")
+            sage: A3 = AtomicSpecies("X, Y")
+            sage: A4 = AtomicSpecies(["X", "Y"])
+            sage: A1 is A2
+            False
+            sage: A3 is A4
+            True
+        """
+        from sage.structure.category_object import normalize_names
+        names = normalize_names(-1, names)
+        return super().__classcall__(cls, names)
+
+    def __init__(self, names):
+        r"""
+        Infinite set of multivariate atomic species.
+
+        INPUT:
+
+        - ``names`` -- an iterable of ``k`` strings
+
+        TESTS::
+
+            sage: At1 = AtomicSpecies(["X"])
+            sage: At2 = AtomicSpecies(["X", "Y"])
+            sage: TestSuite(At1).run(skip="_test_graded_components")
+            sage: TestSuite(At2).run(skip="_test_graded_components")
+        """
+        category = SetsWithGrading().Infinite()
+        Parent.__init__(self, names=names, category=category)
+        ElementCache.__init__(self)
+        self._k = len(names)
+        self._grading_set = IntegerVectors(length=self._k)
+        self._renamed = set()  # the degrees that have been renamed already
+
+    @cached_method
+    def an_element(self):
+        """
+        Return an element of ``self``.
+
+        TESTS::
+
+            sage: At1 = AtomicSpecies("X")
+            sage: At2 = AtomicSpecies("X, Y")
+            sage: At1.an_element()
+            E_2
+            sage: At2.an_element()
+            {((1,2)(3,4),): (frozenset({1, 2}), frozenset({3, 4}))}
+        """
+        G = PermutationGroup([[(2 * i - 1, 2 * i) for i in range(1, self._k + 1)]])
+        m = {i: [2 * i - 1, 2 * i] for i in range(1, self._k + 1)}
+        return self._element_constructor_(G, m)
+
+    def _element_constructor_(self, G, pi=None):
+        r"""
+        Construct the `k`-variate atomic species with the given data.
+
+        INPUT:
+
+        - ``G`` - an element of ``self`` (in this case pi must be ``None``)
+          or a permutation group.
+        - ``pi`` - a dict mapping sorts to iterables whose union is the domain.
+          If `k=1`, `pi` can be omitted.
+        """
+        if parent(G) == self:
+            if pi is not None:
+                raise ValueError("cannot reassign sorts to an atomic species")
+            return G
+        if not isinstance(G, PermutationGroup_generic):
+            raise ValueError(f"{G} must be a permutation group")
+        if pi is None:
+            if self._k == 1:
+                pi = {1: G.domain()}
```

Comment:
Possibly it would be more "pythonic" to index the sorts beginning with `0`.

Reply:
True, but I don't think everything needs to be pythonic. I think it makes more sense from a user standpoint to use 1-based indexing for sorts.

Reply:
Well, almost everything in sage is 0-based, as far as I know.  Notable exceptions are the domain of a permutation group and the labels of a graph.  However, these feel somwhat different - the sorts are more like variables rather than labels.